File: java-algo/src/main/java/com/leetcode/Problem623.java
package com.leetcode;

import com.common.TreeNode;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * 623. Add One Row to Tree
 * https://leetcode.com/problems/add-one-row-to-tree/
 */
public class Problem623 {
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        if (depth == 1) {
            return new TreeNode(val, root, null);
        }
        int d = 1;
        Deque<TreeNode> q = new ArrayDeque<>();
        q.offerLast(root);
        while (!q.isEmpty()) {
            d++;
            int n = q.size();
            for (int i = 0; i < n; i++) {
                TreeNode node = q.pollFirst();
                if (d < depth) {
                    if (node.left != null) {
                        q.offerLast(node.left);
                    }
                    if (node.right != null) {
                        q.offerLast(node.right);
                    }
                } else {
                    node.left = new TreeNode(val, node.left, null);
                    node.right = new TreeNode(val, null, node.right);
                }
            }
        }
        return root;
    }
}

File: python-algo/leetcode/problem_623.py
"""623. Add One Row to Tree
https://leetcode.com/problems/add-one-row-to-tree/
"""
from typing import Optional

from common.tree_node import TreeNode


class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        if depth == 1:
            return TreeNode(val, left=root)
        if not root:
            return None
        k = 1
        q = [root]
        while q:
            cur_level = q[:]
            next_level = []
            while q:
                node = q.pop(0)
                if node.left:
                    next_level.append(node.left)
                if node.right:
                    next_level.append(node.right)
            if k == depth - 1:
                for node in cur_level:
                    left = node.left
                    node.left = TreeNode(val)
                    node.left.left = left
                    right = node.right
                    node.right = TreeNode(val)
                    node.right.right = right
            else:
                q = next_level
            k += 1
        return root
